okhttp/src/main/java/okhttp3/internal/concurrent/TaskQueue.kt

```diff
@@ -39,12 +62,126 @@ interface TaskQueue {
    * is running when that time is reached, that task is allowed to complete before this task is
    * started. Similarly the task will be delayed if the host lacks compute resources.
    */
-  fun schedule(task: Task, delayNanos: Long = 0L)
+  fun schedule(task: Task, delayNanos: Long) {
+    task.initQueue(this)
+
+    synchronized(taskRunner) {
+      if (scheduleAndDecide(task, delayNanos)) {
+        taskRunner.kickCoordinator(this)
+      }
+    }
+  }
+
+  /** Adds [task] to run in [delayNanos]. Returns true if the coordinator should run. */
+  private fun scheduleAndDecide(task: Task, delayNanos: Long): Boolean {
+    val now = taskRunner.backend.nanoTime()
+    val executeNanoTime = now + delayNanos
+
+    // If the task is already scheduled, take the earlier of the two times.
+    val existingIndex = futureTasks.indexOf(task)
```

Comment:
If you trusted task.nextExecuteTime then presumably you could simplify this by using a java PriorityQueue?  

null to x = insert
x to <x = remove and insert
x to >x = don't change nextExecuteTime or modify queue

Reply:
Great idea. Will do.

Reply:
... upon actually trying it, it’s unsatisfying because of the above `now` thing. I’m going to leave it as lists for now, unsatisfying as that is.

Reply:
Thanks for humoring me.